### PR TITLE
Support "merge upstream branch" (Sync fork)

### PR DESCRIPTION
**Pull request by wxiaoguang**
Add basic "sync fork" support (GitHub-like)

<details>

image

</details>

### CODE REVIEW

**Comment by yp05327** [comment]
Great! My favorite feature! 😆
Hope this can be merged into 1.23.

File: templates/repo/code/upstream_diverging_info.tmpl [hunk 1]
```diff
@@ -0,0 +1,12 @@
+{{if and .UpstreamDivergingInfo .UpstreamDivergingInfo.CommitsBehind}}
+<div class="ui message flex-text-block">
+	<div class="tw-flex-1">
+		{{$upstreamLink := printf "%s/src/branch/%s" .Repository.BaseRepo.Link .BranchName}}
+		{{$upstreamHtml := HTMLFormat `<a href="%s">%s:%s</a>` $upstreamLink .Repository.BaseRepo.FullName .BranchName}}
+		{{ctx.Locale.TrN .UpstreamDivergingInfo.CommitsBehind "repo.pulls.upstream_diverging_prompt_behind_1" "repo.pulls.upstream_diverging_prompt_behind_n" .UpstreamDivergingInfo.CommitsBehind $upstreamHtml}}
+	</div>
+	<button class="ui compact green button tw-m-0 link-action" data-url="{{.Repository.Link}}/branches/merge-upstream?branch={{.BranchName}}">
+		{{ctx.Locale.Tr "repo.pulls.upstream_diverging_merge"}}
+	</button>
```
Comment (silverwind) [inline templates/repo/code/upstream_diverging_info.tmpl:15]: Need to hide the button for logged-out users?

File: templates/repo/code/upstream_diverging_info.tmpl [hunk 2]
```diff
@@ -0,0 +1,12 @@
+{{if and .UpstreamDivergingInfo .UpstreamDivergingInfo.CommitsBehind}}
+<div class="ui message flex-text-block">
+	<div class="tw-flex-1">
+		{{$upstreamLink := printf "%s/src/branch/%s" .Repository.BaseRepo.Link .BranchName}}
+		{{$upstreamHtml := HTMLFormat `<a href="%s">%s:%s</a>` $upstreamLink .Repository.BaseRepo.FullName .BranchName}}
+		{{ctx.Locale.TrN .UpstreamDivergingInfo.CommitsBehind "repo.pulls.upstream_diverging_prompt_behind_1" "repo.pulls.upstream_diverging_prompt_behind_n" .UpstreamDivergingInfo.CommitsBehind $upstreamHtml}}
+	</div>
+	<button class="ui compact green button tw-m-0 link-action" data-url="{{.Repository.Link}}/branches/merge-upstream?branch={{.BranchName}}">
```
Comment (silverwind) [inline templates/repo/code/upstream_diverging_info.tmpl:13]: Need to escape branch name or is go template auto escape smart enough?

File: routers/web/repo/view_home.go [hunk 3]
```diff
@@ -338,14 +286,89 @@ func Home(ctx *context.Context) {
...
+	// Get current git entry user currently looking at.
+	entry, err := ctx.Repo.Commit.GetTreeEntryByPath(ctx.Repo.TreePath)
+	if err != nil {
+		HandleGitError(ctx, "Repo.Commit.GetTreeEntryByPath", err)
+		return
+	}
+
+	// prepare the tree path
+	var treeNames, paths []string
+	branchLink := ctx.Repo.RepoLink + "/src/" + ctx.Repo.BranchNameSubURL()
```
Comment (silverwind) [inline routers/web/repo/view_home.go:331]: Does ctx.Repo.BranchNameSubURL() need escaping?

File: templates/repo/code/upstream_diverging_info.tmpl (hunk 1 above)
Comment (delvh) [reply templates/repo/code/upstream_diverging_info.tmpl:15]: Not just logged out users, also users who only have read permission on the repo

File: routers/web/repo/view_home.go (hunk 3 above)
Comment (wxiaoguang) [reply routers/web/repo/view_home.go:331]: No need to change, it only moves old code

File: templates/repo/code/upstream_diverging_info.tmpl (hunk 2 above)
Comment (wxiaoguang) [reply templates/repo/code/upstream_diverging_info.tmpl:13]: It escapes correctly when the branch name is `a%b`

image

File: templates/repo/code/upstream_diverging_info.tmpl (hunk 1 above)
Comment (wxiaoguang) [reply templates/repo/code/upstream_diverging_info.tmpl:15]: Oh yes, forgot to hide the button on UI, done in fc832132c992e54b33490b987fba2e237ec31047

File: routers/web/repo/view_home.go (hunk 3 above)
Comment (wxiaoguang) [reply routers/web/repo/view_home.go:331]: And `BranchNameSubURL` indeed does right for "URL":

image

File: templates/repo/code/upstream_diverging_info.tmpl (hunk 2 above)
Comment (wxiaoguang) [reply templates/repo/code/upstream_diverging_info.tmpl:13]: Hmm, the "printf" one needs to be escaped .... done in e4085c1ef1d5fdd7b7bcc0c8cc0cb35295837851

**Review by silverwind — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Pull request merged**